### PR TITLE
Upstream backport differ script, that works with Github PRs

### PR DESCRIPTION
**Pull request by szilard-nemeth**

### CODE REVIEW

File: github_us_backport_diff_generator.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    except OSError as e:
+        if log_exception:
+            LOG.exception("Failed to create dirs", exc_info=True)
+        # If Errno is File exists, don't raise Exception
+        if e.errno != errno.EEXIST:
+            raise
+    return dirname
+
+
+def download_file_old(url):
```
Comment (9uapaw) [inline github_us_backport_diff_generator.py]: We can safely delete this I presume

File: github_us_backport_diff_generator.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        # Determine jira ID from patch file
+        jira_id, paths = extract_jira_and_paths_from_patch_file(pr_id, jira.patch)
+        jira_ids.add(jira_id)
+        if len(jira_ids) > 1:
+            raise ValueError("Expected Github PRs to belong to one single jira. Multiple Jira IDs found: %s".format(jira_ids))
+
+    if not jira_id:
+        raise ValueError("Jira ID not found!")
+
+    diff_pairs = [(pr_ids[0], i) for i in pr_ids[1:]]
```
Comment (9uapaw) [inline github_us_backport_diff_generator.py:127]: I think this is nicer with itertools combination.
https://www.geeksforgeeks.org/permutation-and-combination-in-python/

File: github_us_backport_diff_generator.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    print("PR diffs will be created for: {}".format(diff_pairs))
+    diff_files = [(pr_id_to_files[pair[0]].diff, pr_id_to_files[pair[1]].diff) for pair in diff_pairs]
```
Comment (9uapaw) [inline github_us_backport_diff_generator.py]: Destructuring would be cleaner here eg. for left, right in diff_pairs.

File: github_us_backport_diff_generator.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    for files in diff_files:
```
Comment (9uapaw) [inline github_us_backport_diff_generator.py]: Here as well.

**Review by 9uapaw — COMMENTED** [review]
Thank you for the contribution @szilard-nemeth. Can you fix the nits I have written?

File: github_us_backport_diff_generator.py (hunk 1 above)
Comment (szilard-nemeth) [reply github_us_backport_diff_generator.py]: Good point. removed.

File: github_us_backport_diff_generator.py (hunk 2 above)
Comment (szilard-nemeth) [reply github_us_backport_diff_generator.py:127]: The scenario here is not using all combinations of the PRs, I'm pairing the first PR with all other PRs in the list to create the diff html.

See this: 
```
pr_ids = ['1111', '2222', '3333']
diff_pairs = [(pr_ids[0], i) for i in pr_ids[1:]]
diff_pairs
[('1111', '2222'), ('1111', '3333')]

from itertools import combinations
combs = combinations(pr_ids, 2)
list(combs)
[('1111', '2222'), ('1111', '3333'), ('2222', '3333')]
diff_pairs
[('1111', '2222'), ('1111', '3333')]
```
I could use some filtering over itertools.combinations to only accept the pairs where the first item in the tuple is the first item of the pr_ids list, but I don't think it's worth the hassle :)

File: github_us_backport_diff_generator.py (hunk 3 above)
Comment (szilard-nemeth) [reply github_us_backport_diff_generator.py]: good point, fixed.

File: github_us_backport_diff_generator.py (hunk 4 above)
Comment (szilard-nemeth) [reply github_us_backport_diff_generator.py]: fixed.

**Comment by 9uapaw** [comment]
Thanks for the contribution @szilard-nemeth. Latest change looks good to me, merged.

**Comment by szilard-nemeth** [comment]
Thanks @9uapaw

**Pull request merged**